text/1002-required-by-default.rst

```diff
@@ -0,0 +1,178 @@
+not having a true schema. Unless it's necessary to make a field optional,
+forcing fields to be required ensures that whenever data *can* be specified,
+it won't be forgotten at INSERT or UPDATE time.
+
+Required fields simplify type information. There's two particular problems
+on the client side with unexpected optionality:
+
+* lack of handling for that special case in user code, and
+
+* surprising effects of EdgeQL queries that omit objects with unexpected
+  empty sets in fields used in the query.
+
+Downsides
+~~~~~~~~~
+
+* The main disadvantage at time of writing is that the current behavior as of
+  EdgeDB v1.0a3 is the opposite: fields without an explicit qualifier are
+  assumed optional. This poses a migration challenge.
+
+* Required fields have to be always specified at INSERT time and often
+  specified at UPDATE time, making queries more verbose to write and data
+  more costly to send over the wire. This can be dealt with by providing
+  default values with SET DEFAULT which is a form of optionality with an
+  explicit fallback value which doesn't fall outside of the declared type.
+
+* Thrift documentation explains that "required is forever". What they mean
+  by this is that required fields have to be provided by the caller, say a
+  mobile device. Changing a required field into an optional field requires
+  for the database to be updated first before any application code can use
+  this capability. While this is a valid concern for RPC systems where new
+  clients won't be able to connect to old servers, in the case of central
+  databases, the most popular deployment scheme already is to migrate the
+  database first. The central base is an easier target to control. It's true
+  that there's a risk that this ordering will not be kept if the schema
+  definition files are shared between teams responsible for backend and
+  frontend code.
+
+
+Rejected ideas
+==============
+
+Assume OPTIONAL qualifiers when no explicit qualifier is provided
+-----------------------------------------------------------------
+
+This is the current behavior.
+
+An optional field is a form of incomplete data that allows the application
+to flexibly decide how to proceed. If the data is there, fine. If it isn't,
+the application can include user-level code to deal with that.
+
+Plenty of examples of existing databases and RPC systems assume optionality
+by default. Protocol buffers removed support for required fields altogether
+in version 3. JSON-based NoSQL databases like MongoDB assume optionality.
+HTML form fields are optional by default.
+
+Downsides
+~~~~~~~~~
+
+Developers might be failing to fill fields at INSERT time due
+to forgetfulness at query creation time, even if the data could be
+provided. They might be harder to fill at a later stage.
+
+Optionality is a special case that the user code has to deal with. Empty sets
+in queries might leave to unexpected results, i.e. missing objects. An empty
```

Comment:
This is a double-edged sword: https://github.com/edgedb/rfcs/pull/12/files#r441843313